roachprod: create MIG clusters on GCE

herkolategan opened the pull request:
Previously, cluster creation only supported creating VMs individually that were loosely tied by name and labels. This change introduces the ability to pass '--gce-managed' when creating clusters using the Google Cloud provider.

Setting the `managed` flag changes the way a cluster is created in a couple of ways. First an instance template is created that will serve as the template for all VMs that will be created as part of the cluster. The template is specific to the cluster and named as such. After template creation a managed instance group (MIG) is created for each zone requested.

The instance groups, in each region, are then instructed to create the desired number of VMs. Roachprod will then also wait for each instance group to reach a stable state before proceeding.

See: #118207
Epic: CRDB-33832

Release Note: None

herkolategan committed:
roachprod: create MIG clusters on GCE

Previously, cluster creation only supported creating VMs individually that were
loosely tied by name and labels. This change introduces the ability to pass
'--gce-managed' when creating clusters using the Google Cloud provider.

Setting the `managed` flag changes the way a cluster is created in a couple of
ways. First an instance template is created that will serve as the template for
all VMs that will be created as part of the cluster. The template is specific to
the cluster and named as such. After template creation a managed instance group
(MIG) is created for each zone requested.

The instance groups, in each region, are then instructed to create the desired
number of VMs. Roachprod will then also wait for each instance group to reach a
stable state before proceeding.

See: #118207
Epic: CRDB-33832

Release Note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118423)
<!-- Reviewable:end -->

herkolategan committed:
roachprod: destroy MIG resources

After adding MIG support, the way resources are destroyed for `managed` clusters
need to be updated. Instead of deleting the instances, as was done previously,
we destroy the MIG that will destroy the VMs as well. Only after the MIGs have
been destroyed we can also delete the instance template originally created for
the cluster.

Resolves: #118419
Epic: CRDB-33832

Release Note: None

renatolabs commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1178,74 +1223,227 @@ func (p *Provider) Create(
-		argsWithZone = append(argsWithZone, zoneHosts...)
+// createInstanceGroups creates an instance group for the cluster in each zone
+// specified. Instance groups will be created for the specified zones even if no
+// instances are created in that zone during cluster creation.
```
> Instance groups will be created for the specified zones even if no instances are created in that zone during cluster creation.

Why is this the case? If I'm reading the code right, we're only passing `zones` that were actually used when creating the cluster.

renatolabs commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1178,74 +1223,227 @@ func (p *Provider) Create(
+		"--template", templateName,
+		"--size", "0",
+		"--stateful-external-ip", "enabled",
+		"--stateful-internal-ip", "enabled",
```
Why do we need the IPs to be stateful? Might be good to include this in the comments too.

renatolabs commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1178,74 +1223,227 @@ func (p *Provider) Create(
+	// Determine number of stateful disks the instance group should retain.
+	numStatefulDisks := 1
+	if !opts.SSDOpts.UseLocalSSD {
+		numStatefulDisks = 2
```
I don't understand this logic.

renatolabs commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1178,74 +1223,227 @@ func (p *Provider) Create(
+			for _, host := range zoneHosts {
+				argsWithHost := append(argsWithZone[:len(argsWithZone):len(argsWithZone)], []string{"--instance", host}...)
+				g.Go(func() error {
+					fmt.Printf("Command to run: %s\n", argsWithHost)
```
Did you mean to keep this?

renatolabs commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1357,8 +1555,123 @@ func propagateDiskLabels(
+// managed or not at all.
+func isManaged(vms vm.List) bool {
+	firstVM := vms[0]
+	for k, v := range firstVM.Labels {
```
It's a map, so don't need to linear search. I suggest `return vms[0].Labels[ManagedLabel] == "true"`.

renatolabs commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1357,8 +1555,123 @@ func propagateDiskLabels(
+func (p *Provider) deleteManaged(l *logger.Logger, vms vm.List) error {
+	projectClusterMap := make(map[string]map[string]struct{})
+	for _, v := range vms {
+		if v.Provider != ProviderName {
```
IMO, this is a little too defensive. The other cloud-specific functions seem to assume that it's already part of the contract that the `vms` passed are from this provider.

renatolabs commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1357,8 +1555,123 @@ func propagateDiskLabels(
+		if err != nil {
+			return err
+		}
+		if projectClusterMap[v.Project] == nil {
```
Do we support _cross-project_ clusters? I don't think so. If not, we can simplify this code.

renatolabs reviewed (commented):
Very exciting! 🎉

renatolabs commented:
A usability note: I was playing around with managed instances on this branch, and the first time I tried to create a cluster, it failed because I was running an older version of `gcloud` that didn't recognize the `--stateful-*` flags. Upgrading gcloud fixed it.

However, it got me into a state where I couldn't `roachprod destroy` the cluster because the cluster hadn't been "created", and I also couldn't just run the same `create` command again because the instance template already existed. So I had to manually delete it (in the Google Cloud console) and run `create` again.

Given that this flag was promoted to GA relatively recently[^1], it might be a good idea to do a version check before attempting to create the cluster.

[^1]: October 2023, it seems: https://groups.google.com/g/google-cloud-sdk-announce/c/PmuWoyRNzck.

herkolategan commented:
> However, it got me into a state where I couldn't `roachprod destroy`

Good catch, I thought the template might cause this kind of half-created state issue, but it slipped off my radar at some point. One issue with this is you need at least one VM to be able to determine the project in order to list the templates. If you have one VM already, there is definitely a MIG so the post template deletion will then also always work. The issue comes in if MIG creation fails completely. Maybe the best way to handle this is to do deletion in the create command then 🤔 ? But I'll think about this a bit more, maybe the delete API needs updating to handle cases where there are no VMs.

> Given that this flag was promoted to GA relatively recently[1](https://github.com/cockroachdb/cockroach/pull/118423#user-content-fn-1-fd78371c167ea42616c8eca91ecc7576), it might be a good idea to do a version check before attempting to create the cluster.

Good point, I'll add this.

herkolategan reviewed (commented):
Thanks for the thorough review! A good part of this code is just relocating existing code and subdividing it into new functions, but it's good to bring in some improvements while we are at it.

The bigger concern here is the half-created state that is not resolvable from `roachprod` itself. Ideally I would like this to be full-proof, but it might need some more intensive changes. One silver lining is that in that state it doesn't cost anything, but it's not great to be stuck trying to create a cluster.

herkolategan committed:
roachprod: destroy empty MIG cluster resources

Previously, if a MIG failed to create or for some other reason we had either
empty MIGs or just an instance template these resources would cause a scenario
where the cluster can neither be deleted nor recreated.

In this change we update the gcloud provider's `List` to add an Empty Cluster
when an instance template exists, but there are no VMs present. This allows the
`Delete` command to find both the template or any empty MIGs and clean up the
resources correctly. A MIG requires an instance template to exist so the case
would cover both scenarios.

See: #118419
Epic: CRDB-33832

Release Note: None

herkolategan committed:
roachprod: verify gcloud version for MIGs

Previously, creating a managed (MIG) cluster in GCE could fail due to some of
the required arguments not begin supported by lower `gcloud` versions. This
change adds a minimum required version check when creating clusters with the
managed flag.

Epic: CRDB-33832

Release Note: None

herkolategan commented:
@DarrylWong Added you as reviewer just to have a pass over the `Empty Cluster` logic, and if I'm not missing anything regarding it.

renatolabs reviewed (approved):
LGTM!

Not sure if you have done it yet, but it would be nice to do a `gc` run manually to make sure we are dealing with partially created clusters as we expect.

DarrylWong commented:
I'm still trying to wrap my head around the issue at hand. In the example Renato mentioned, what actually exists here? Just the template but not the actual instance group? If so I want to point out that the only reason the "fake VM" approach works for azure is that for azure, we  delete the entire cluster vs each vm one by one like we do in gce:

https://github.com/cockroachdb/cockroach/blob/c8c0999464cfd616ebd389515119175e8991f3c7/pkg/roachprod/cloud/cluster_cloud.go#L291-L294

This is a key difference as `deleteClusters` retrieves the resource group from the fake VM and deletes that. If we try to call `delete` on the fake VM, it will fail because that VM doesn't actually exist. This is why we don't use the `badVMs` logic to delete empty clusters in GC since that calls `delete`.

https://github.com/cockroachdb/cockroach/blob/8261c78a3bffbb2ff6b0418f0d9230453220f21f/pkg/roachprod/vm/gce/gcloud.go#L1673-L1674

So I think deleting a non existent instance group would fail. At least in Azure, maybe GCE handles it differently though?

DarrylWong reviewed (approved):
Hmm actually scratch that 😅, I missed the key bit about `listManagedInstanceGroups` getting it's list by querying gcloud and not through `vms`. In that case it LGTM.

herkolategan commented:
> This is a key difference as `deleteClusters` retrieves the resource group from the fake VM and deletes that. If we try to call `delete` on the fake VM, it will fail because that VM doesn't actually exist. This is why we don't use the `badVMs` logic to delete empty clusters in GC since that calls `delete`.

Ah yes, I picked up on that, and instead of `deleteClusters` in Azure, the distinction is on `managed` clusters in GCE. Only managed clusters / MIGs will ever use the empty VM path, which only need a cluste rname vs. actual VMs. So it's not possible for it to end up in regular `Delete`. Although you have made me thought about the fact that this is never stated explicitly somewhere, and rather only visible through a good look at the code and paths it follows in those scenarios. The thought that we need a way to associate other resources with VMs & clusters somehow in our Provider APIs did cross my mind as we're starting to introduce features that the API is struggling to carry as it is.

herkolategan commented:
TFTRs!

bors r=renatolabs,DarrylWong

srosenberg commented:
Catching up on a belated review... Nice work! We have MIGs 🚀
